.github/workflows/codeql-analysis.yml

```diff
@@ -0,0 +1,32 @@
+name: "CodeQL"
+# see https://codeql.github.com/docs/
+
+on:
+  push:
+    branches: [master]
+  pull_request:
+    # The branches below must be a subset of the branches above
+    branches: [master]
+
```

Comment:
> The analyzer is now working for basic builds. We just need the found issues to be resolved before this can merge.

Excellent!

Now that I know a lot more about GitHub Actions, I will do another scan of the proposed static analysis configuration, but one (future) change request seems likely already: Once #1160 lands, we will probably want to merge the static analysis job into `default.yaml` being added by that PR. In my tests, having two workflows creates more noise/confusion than one workflow with two jobs. If we can unify the `branches:` configuration, then we should probably merge the two configurations into one, adding a CodeQL job to `default.yaml`.

Reply:
What do you mean by noise/confusion exactly?

Ideally we want ~3 workflows since the tasks vary relevance by branch and event:
 * maintenance - spelling, syntax style enforcement, etc
 * source testing - test-builds.sh, cppunit tests and static analysis
 * black-box testing - daft protocol checks etc.

Reply:
> > Alex: we will probably want to merge the static analysis job into default.yaml being added by that PR. In my tests, having two workflows creates more noise/confusion than one workflow with two jobs

> Amos: What do you mean by noise/confusion exactly?

I think I observed several negative side effects, but the one I remember well is the loss of clear/obvious correlation between a PR and its Actions/tests on the Actions [page](https://github.com/squid-cache/squid/actions): With multiple workflows, it becomes difficult for a human to quickly see which list entries are the latest/current results for a given PR. That list contains old/stale and current/fresh entries. A human can easily look for the latest entry with the familiar PR title (or number), but that single entry is _not_ what they need to look for when there are multiple workflows -- they need to find multiple latest list entries, one for each workflow.
 
> Ideally we want ~3 workflows since the tasks vary relevance by branch and event:

I disagree. Ideally, we want the smallest number of workflows that still correctly test what we want to test. That number may be one, two, three, or more (depending on our evolving testing needs that should be assessed separately), but the ideal is the minimal sufficient number of workflows because each workflow adds overheads, especially given current Actions/YAML difficulties with supporting code reuse.

> * maintenance - spelling, syntax style enforcement, etc
> * source testing - test-builds.sh, cppunit tests and static analysis
> * black-box testing - daft protocol checks etc.

You are describing three kinds of tests, not necessarily three kinds of workflows. IMO, the correct number of workflows is determined by unique combinations of YAML fields like `on:` and `env:`. For example, if we can keep all such fields the same for all three groups of tests mentioned above, then we should have just one workflow file (with three or more jobs, one for each test).

Reply:
Several change requests depend on whether we merge the two workflows into one file. Let's use _this_ existing thread to make that decision.

Reply:
Okay, I will merge with default.yaml.